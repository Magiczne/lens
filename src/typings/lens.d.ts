import { LoadEvent, Viewport } from 'puppeteer'
import { Logger } from '@/typings/logging'
import { RulesetParser, RulesetValidator } from '@/typings/rules'

interface ArgumentParser {
    parse (args: LensArguments): ParsedLensArguments

    parseDirectory (dir?: string): string | undefined
    parseResolution (resolution?: string): Record<string, Viewport[]>
    parseUrl (url: string): URL[]
}

interface LensArguments {
    [x: string]: unknown

    url?: string
    resolution?: string
    tag: string

    inputDir?: string
    outputDir?: string

    _: string[]
    $0: string
}

interface ParsedLensArguments {
    urls: URL[]
    resolutions: Record<string, Viewport[]>
    tag: string

    inputDir?: string
    outputDir?: string
}

interface LensConfig {
    chunkSize: number
    directories: {
        input: string
        output: string
    }
    puppeteer: {
        headless: boolean
        waitUntil: LoadEvent | LoadEvent[]
    }
}

interface LensDependencies {
    argumentParser: ArgumentParser
    logger: Logger
    rulesetParser: RulesetParser
    rulesetValidator: RulesetValidator
}

export {
    ArgumentParser,
    LensArguments, ParsedLensArguments,
    LensConfig, LensDependencies
}
