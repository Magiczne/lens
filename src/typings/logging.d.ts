interface Logger {
    error (message: string): void
    header (message: string): void
    info (message: string): void
    success (message: string): void
}

export {
    Logger
}
