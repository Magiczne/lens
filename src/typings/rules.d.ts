import { Page, Viewport } from 'puppeteer'

interface Rule {
    url: string
    tag: string
    renderFor?: Array<string> | Array<Viewport>

    afterPageLoaded?: (page: Page) => Promise<void> | void
}

interface ParsedRule {
    url: URL
    tag: string
    renderFor: Record<string, Array<Viewport>>

    afterPageLoaded?: (page: Page) => Promise<void> | void
}

interface Ruleset {
    disable: boolean
    rules: Array<Rule>
}

interface ParsedRuleset {
    disable: boolean
    rules: Array<ParsedRule>
}

interface RulesetParser {
    parse (rawRuleset: Ruleset): ParsedRuleset

    parseRule (rule: Rule): ParsedRule
}

interface RulesetValidator {
    validate (ruleset: Record<string, unknown>, file: string): Promise<Ruleset> | Ruleset
}

export {
    Rule, Ruleset,
    ParsedRule, ParsedRuleset,
    RulesetParser, RulesetValidator
}
