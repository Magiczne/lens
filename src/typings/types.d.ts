import { UrlWithStringQuery } from 'url'

interface Resolution {
	width: number
	height: number
}

interface LensArguments {
	[x: string]: unknown

	url: string
	resolution?: string
	tag: string

	_: string[]
	$0: string
}

interface ParsedLensArguments {
	urls: UrlWithStringQuery[]
	resolutions?: Resolution[]
	tag: string
}

interface Logger {
	error (message: string): void
	header (message: string): void
	info (message: string): void
	success (message: string): void
}

export {
	LensArguments, ParsedLensArguments, Resolution,
	Logger
}
