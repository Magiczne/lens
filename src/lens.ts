import fs from 'fs'
import puppeteer, { Browser } from 'puppeteer'
import url, { UrlWithStringQuery } from 'url'

import { LensArguments, Logger, ParsedLensArguments, Resolution } from './typings/types'
import { ConsoleLogger } from './logger'
import { forEachAsync } from './utils'

export default class Lens {
	private readonly screenshotsDir = './screenshots'

	private readonly args: ParsedLensArguments
	private browser: Browser | undefined
	private logger: Logger

	public constructor (args: LensArguments, logger: Logger | undefined = undefined) {
		this.args = this.parseArguments(args)
		this.logger = logger ?? new ConsoleLogger()
	}

	/**
	 * Create directory for screenshots if it does not exist,
	 * then instantiate browser.
	 */
	public async init (): Promise<void> {
		if (!fs.existsSync(this.screenshotsDir)) {
			try {
				fs.mkdirSync(this.screenshotsDir, { recursive: true })
				this.logger.info(`Created ${this.screenshotsDir}`)
			} catch (e) {
				this.logger.error(`Could not create screenshots directory`)

				process.exit(14)
			}
		}

		this.browser = await puppeteer.launch()
	}

	/**
	 * Run lens
	 */
	public async run (): Promise<void> {
		if (!this.browser) {
			this.logger.error('Lens has not been initialized. Please run "init" before running.')

			process.exit(15)
		}

		await forEachAsync(this.args.urls, async u => {
			this.logger.header(`Running lens for ${u.href}`)

			const directory = this.createDirectoryForUrl(u, this.args.tag)
			await this.generateScreenshots(this.args, u, directory)
		})
	}

	/**
	 * Create directory where screenshots of the specified url will be stored
	 *
	 * @param url
	 * @param tag
	 * @private
	 */
	private createDirectoryForUrl (url: UrlWithStringQuery, tag = ''): string {
		let directory = `${this.screenshotsDir}/${url.host}`
		if (tag) {
			directory = `${directory}/${tag}`
		}

		if (!fs.existsSync(directory)) {
			try {
				fs.mkdirSync(directory, { recursive: true })
				this.logger.info(`Created ${directory}`)
			} catch {
				this.logger.error(`Could not create directory ${directory}`)

				process.exit(14)
			}
		}

		return directory
	}

	/**
	 * Generate screenshots based on the specified arguments
	 *
	 * @param args
	 * @param url
	 * @param dir
	 * @private
	 */
	private async generateScreenshots (args: ParsedLensArguments, url: UrlWithStringQuery, dir: string): Promise<void> {
		await Promise.all(args.resolutions.map(async res => {
			if (!this.browser) return // TODO: Handle that in a prettier way

			const page = await this.browser.newPage()

			await page.setViewport({ ...res })
			await page.goto(url.href)
			await page.screenshot({
				path: `${dir}/${res.width}x${res.height}.png`
			})

			await page.close()

			this.logger.success(`[DONE] ${url.host} ${res.width}x${res.height}`)
		}))
	}

	/**
	 * Do the cleanup
	 */
	public async dispose (): Promise<void> {
		await this.browser?.close()
	}

	/**
	 * Parse arguments from user and return them
	 *
	 * @param args
	 * @private
	 */
	private parseArguments (args: LensArguments): ParsedLensArguments {
		const urls: UrlWithStringQuery[] = args.url.split(' ')
			.map(u => url.parse(u))

		const resolutions: Resolution[] = args.resolution
			? args.resolution.split(' ')
				.map(res => {
					return res.trim()
						.split('x')
						.map(x => parseInt(x, 10))
				})
				.map(res => {
					return {
						width: res[0],
						height: res[1]
					}
				})
			: []

		return {
			urls: urls,
			resolutions: resolutions,
			tag: args.tag,
		}
	}
}
