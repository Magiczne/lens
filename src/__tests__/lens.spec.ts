import path from 'path'import Lens from '@/lens'import ConsoleLogger from '@/logging/console-logger'import DefaultArgumentParser from '@/parsing/argument-parser'import DefaultRulesetParser from '@/parsing/ruleset-parser'import { LensArguments, LensConfig, LensDependencies, ParsedLensArguments } from '@/typings/lens'import DefaultRulesetValidator from '@/validation/ruleset-validator'jest.mock('@/logging/console-logger')jest.mock('@/parsing/argument-parser')jest.mock('@/parsing/ruleset-parser')jest.mock('@/validation/ruleset-validator')describe('Lens', () => {    const defaultDependencies: LensDependencies = {        argumentParser: new DefaultArgumentParser(),        logger: new ConsoleLogger(),        rulesetParser: new DefaultRulesetParser(),        rulesetValidator: new DefaultRulesetValidator()    }    const defaultArgs: LensArguments = {        _: [],        $0: '',        url: ['https://example.com'],        resolution: ['1280x720'],        viewports: ['desktop', 'tablet', 'phone'],        tag: 'default'    }    const defaultConfig: LensConfig = {        chunkSize: 5,        directories: {            input: path.resolve('./rules'),            output: path.resolve('./screenshots')        },        puppeteer: {            headless: true,            waitUntil: [                'load', 'networkidle2'            ]        }    }    const lensFactory = (        dependencies?: Partial<LensDependencies>,        args?: Partial<LensArguments>,        config?: Partial<LensConfig>    ): Lens => {        return new Lens({            ...defaultDependencies,            ...dependencies        }, {            ...defaultArgs,            ...args        }, {            ...defaultConfig,            ...config        })    }    describe('run', (): void => {        it('should throw critical error when run without initialization', (): void => {            const lens = lensFactory()            expect(lens.run())                .rejects                .toThrowErrorMatchingSnapshot()        })        it('should run from args when url was passed in args', async (): Promise<void> => {            const parser = new DefaultArgumentParser()            jest.spyOn(parser, 'parse').mockImplementation((): ParsedLensArguments => {                return {                    urls: [                        new URL('https://example.com')                    ]                } as ParsedLensArguments            })            const lens = lensFactory({ argumentParser: parser })            const runFromArgsSpy = jest.spyOn(lens, 'runFromArgs').mockImplementation()            const runFromRulesetSpy = jest.spyOn(lens, 'runFromRuleset').mockImplementation()            await lens.init()            await lens.run()            expect(runFromArgsSpy).toBeCalledTimes(1)            expect(runFromRulesetSpy).not.toBeCalled()            await lens.dispose()        })        it('should run from ruleset when url was not passed in args', async (): Promise<void> => {            const lens = lensFactory()            const runFromArgsSpy = jest.spyOn(lens, 'runFromArgs').mockImplementation()            const runFromRulesetSpy = jest.spyOn(lens, 'runFromRuleset').mockImplementation()            await lens.init()            await lens.run()            expect(runFromArgsSpy).not.toBeCalled()            expect(runFromRulesetSpy).toBeCalledTimes(1)            await lens.dispose()        })    })})