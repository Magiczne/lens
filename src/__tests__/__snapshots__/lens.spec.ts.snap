// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lens run should throw critical error when run without initialization 1`] = `"Lens has not been initialized. Please run \\"init\\" before running."`;
